safemysql.class.php

```diff
@@ -170,6 +174,39 @@ public function query()
 		return $this->rawQuery($this->prepareQuery(func_get_args()));
 	}
 
+
+	/**
+	 * Run the provided callback inside a transaction
+	 * @param  callable $callback
+	 * @param  [boolean] $can_retry If the whole transaction should be retried, if it fails due to a deadlock
+	 * @return mixed whatever the callback returns
+	 */
+	public function transaction($callback, $can_retry = false)
```

Comment:
- [ ] Hmm. There's an interesting bug here (or at least a case that we're failing to handle cleanly) that also existed in our internal `MysqlUtils` implementation. We *ought* to be keeping track of whether we're already in a transaction (with an instance variable) and throwing an error if somebody tries to call `transaction()` while already in a transaction. Based upon a quick experiment, MySQL's behaviour when calling `START TRANSACTION` while already in a transaction seems to be pretty unhelpful; you'd expect an error, but instead it *commits* the current transaction(!!!) and starts a new one.